### PR TITLE
[JSON source gen 3/3] Add new methods to JsonSerializer and System.Net.Http.Json APIs that take type metadata

### PR DESCRIPTION
**Pull request by layomia**
Contributes to https://github.com/dotnet/runtime/issues/45448.

PR 1 - https://github.com/dotnet/runtime/pull/51149
PR 2 - https://github.com/dotnet/runtime/pull/51300
PR 3 - this PR

This PR completes the initial implementation of the source generator needed for community feedback (preview 4) and for Blazor to adopt the new pattern. cc @CoffeeFlux, @eerhardt, @ericstj, @Anipik.

https://github.com/dotnet/runtime/issues/51544 tracks the remaining S.N.H.J APIs we need to add.

cc @Jozkee for System.Net.Http.Json changes.

### CODE REVIEW

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs [hunk 1]
```diff
@@ -608,32 +622,32 @@ public partial struct JsonNodeOptions
...
         System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw null; }
         public bool TryGetPropertyValue(string propertyName, out System.Text.Json.Node.JsonNode? jsonNode) { throw null; }
         public override void WriteTo(System.Text.Json.Utf8JsonWriter writer, System.Text.Json.JsonSerializerOptions? options = null) { }
     }
     public abstract partial class JsonValue : System.Text.Json.Node.JsonNode
     {
-        private protected JsonValue(JsonNodeOptions? options = null) { throw null; }
-        public static System.Text.Json.Node.JsonValue? Create<T>(T value, System.Text.Json.Node.JsonNodeOptions? options = default(System.Text.Json.Node.JsonNodeOptions?)) { throw null; }
-        public abstract bool TryGetValue<[System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute(System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.PublicConstructors | System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.PublicFields | System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.PublicProperties)]T>([System.Diagnostics.CodeAnalysis.NotNullWhen(true)] out T? value);
+        internal JsonValue() { }
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/ref/System.Text.Json.cs:648]: Was this change intended?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs [hunk 2]
```diff
@@ -29,7 +29,7 @@ public sealed partial class JsonSerializerOptions
 
         internal JsonSerializerContext? _context;
 
-        private Func<Type, JsonSerializerOptions, JsonTypeInfo>? _typeInfoCreationFunc = null!;
+        private static Func<Type, JsonSerializerOptions, JsonTypeInfo>? _typeInfoCreationFunc = null!;
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:32]: Are we sure this should be static? How does this interact when I have 2 Options instances, one from source gen and one not from source gen. If I try to serialize something from source gen before using the non-source gen options, it will throw an exception. But if the ordering gets changed, suddenly the same operation will stop throwing because the non-source gen options affected static state.

If this should really be static, the field should be named `s_typeInfoCreationFunc`.

Also, what is the value in saying this isn't nullable, but then we check it for null later in the code?

File: src/libraries/System.Net.Http.Json/ref/System.Net.Http.Json.cs [hunk 3]
```diff
@@ -9,12 +9,16 @@ namespace System.Net.Http.Json
...
         public static System.Threading.Tasks.Task<TValue?> GetFromJsonAsync<TValue>(this System.Net.Http.HttpClient client, string? requestUri, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public static System.Threading.Tasks.Task<TValue?> GetFromJsonAsync<TValue>(this System.Net.Http.HttpClient client, System.Uri? requestUri, System.Text.Json.JsonSerializerOptions? options, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public static System.Threading.Tasks.Task<TValue?> GetFromJsonAsync<TValue>(this System.Net.Http.HttpClient client, System.Uri? requestUri, System.Text.Json.Serialization.Metadata.JsonTypeInfo<TValue> jsonTypeInfo, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Http.Json/ref/System.Net.Http.Json.cs:21]: Don't we need equivalent overloads for the `Post` and `Put` methods as well?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs (hunk 2 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:32]: What is the downside of not having an exception (i.e. having things work) in the case where you used source-gen overloads first? Chances are you shouldn't even hit the code-path where we could invoke this func. If all root serializable types are declared with `JsonSerializableAttribute`, along with any types that could be serialized polymorphically, then you'll never hit the code-path where we try to create the metadata using reflection.

One downside is that folks might not configure their source-gen options correctly and wouldn't know because "things just work". Is this the problem we are avoiding by making it an instance field? Folks that care about size and not doing run-time reflection would likely profile their apps to verify.

Making the field static would mean avoiding an extra alloc for each `options` instance. cc @steveharter in case that was the perspective from https://github.com/dotnet/runtime/pull/51300#discussion_r614133566.

---

> Also, what is the value in saying this isn't nullable, but then we check it for null later in the code?

It is nullable - I'll remove the ` = null!` as that is not needed.

File: src/libraries/System.Net.Http.Json/ref/System.Net.Http.Json.cs (hunk 3 above)
Comment (layomia) [reply src/libraries/System.Net.Http.Json/ref/System.Net.Http.Json.cs:21]: I opened https://github.com/dotnet/runtime/issues/51544 to track it. It involves a bit more work than I initially thought (described in that issue).

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:32]: > What is the downside of not having an exception (i.e. having things work) in the case where you used source-gen overloads first?

In general, having static mutable state is bad, for many reasons.

Here's the case I'm concerned about:

Working code:

```C#
// 1. Code uses non-source gen serialization at the beginning of the app
var options = new JsonSerializationOptions();
object someObject = ...;
JsonSerializer.Serialize(someObject, options);

// The above code "initializes" the static variable, causing the below code to "work"

// 2. Somewhere else in the app, code is using the source gen APIs, but they don't have `JsonSerializableAttribute` for a type. The code in (1) above executed first, so this code executes successfully:

JsonSerializer.Serialize(stronglyTypeObject, MyJsonContext);
```

Sometime later, the app changes and no longer needs to do `(1)` above. So that code gets deleted.

No longer working code:

```C#
// This now throws somewhere completely unrelated to removing (1) above. Note that (1) and (2) may be in unrelated referenced libraries. 
JsonSerializer.Serialize(stronglyTypeObject, MyJsonContext);
```

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:32]: Yes that makes sense. One could still expect that people using the source-gen pattern do the right thing, but it will be hard to correct mistakes if they span across different assemblies.

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:648]: No, it was done by the tool. I've reverted this.

File: src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs [hunk 4]
```diff
@@ -37,21 +39,61 @@ public static partial class HttpContentJsonExtensions
 
         private static async Task<object?> ReadFromJsonAsyncCore(HttpContent content, Type type, Encoding? sourceEncoding, JsonSerializerOptions? options, CancellationToken cancellationToken)
         {
-            Stream contentStream = await ReadHttpContentStreamAsync(content, cancellationToken).ConfigureAwait(false);
+            using (Stream contentStream = await GetContentStream(content, sourceEncoding, cancellationToken).ConfigureAwait(false))
+            {
+                return await JsonSerializer.DeserializeAsync(contentStream, type, options ?? JsonContent.s_defaultSerializerOptions, cancellationToken).ConfigureAwait(false);
```
Comment (steveharter) [inline src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs:44]: This is related to using the new `DeserializeAsync()` and not the code-gen effort, correct?

File: src/libraries/System.Net.Http.Json/tests/FunctionalTests/JsonContext/String.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+namespace System.Net.Http.Json.Functional.Tests
+{
+    internal partial class JsonContext : JsonSerializerContext
+    {
+        private JsonTypeInfo<string> _String;
+        public JsonTypeInfo<string> String
+        {
+            get
+            {
+                if (_String == null)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Http.Json/tests/FunctionalTests/JsonContext/String.cs:16]: Nit: this class seems to repeat itself a few times in the test code. Could it perhaps be changed into an abstract base class?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs [hunk 6]
```diff
@@ -13,35 +13,6 @@ public static partial class JsonSerializer
...
-                Debug.Assert(success);
-            }
-            catch
-            {
-                state.DisposePendingDisposablesOnException();
-                throw;
-            }
-        }
-
         private static bool WriteCore<TValue>(
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs:16]: nit: "WriteCore" and "SerializeUsingMetadata" have different terminology. Suggest making them consistent.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/JsonContext/HighLowTemps.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+                options,
+                isProperty: true,
+                declaringType: typeof(HighLowTemps),
+                propertyTypeInfo: jsonContext.Int32,
+                converter: null,
+                getter: static (obj) => { return ((HighLowTemps)obj).High; },
+                setter: static (obj, value) => { ((HighLowTemps)obj).High = value; },
+                ignoreCondition: default,
+                numberHandling: default,
+                propertyName: "High",
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/JsonContext/HighLowTemps.cs:58]: nit (for future): using `nameof(HighLowTemps.High)` is a bit less error-prone and will give a nice compiler error if properties are removed\renamed (but so will the code gen'd for the getter and setter, so not all that important).

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/JsonContext/JsonContext.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+        public JsonContext() : base(null)
+        {
+        }
+
+        public JsonContext(JsonSerializerOptions options) : base(options)
+        {
+        }
+
+        private JsonConverter GetRuntimeProvidedCustomConverter(System.Type type)
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/JsonContext/JsonContext.cs:22]: nit: add "using System;" and change param type to "Type"

File: src/libraries/System.Net.Http.Json/tests/FunctionalTests/JsonContext/Person.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+                            default);
+                    }
+                }
+
+                return _Person;
+            }
+        }
+        private static JsonPropertyInfo[] PersonPropInitFunc(JsonSerializerContext context)
+        {
+            JsonContext jsonContext = (JsonContext)context;
```
Comment (steveharter) [inline src/libraries/System.Net.Http.Json/tests/FunctionalTests/JsonContext/Person.cs:43]: Can this cast ever fail?

**Review by steveharter — APPROVED** [review]
LGTM; some nits and a couple questions.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs [hunk 10]
```diff
@@ -85,7 +86,104 @@ public static partial class JsonSerializer
             if (returnType == null)
                 throw new ArgumentNullException(nameof(returnType));
 
-            return ReadAllAsync<object>(utf8Json, returnType, options, cancellationToken);
+            return ReadAllUsingOptionsAsync<object>(utf8Json, returnType, options, cancellationToken);
+        }
+
+        private static ValueTask<TValue?> ReadAllUsingOptionsAsync<TValue>(
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs:92]: Nit: keeping private methods above public APIs.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs [hunk 11]
```diff
@@ -100,7 +198,7 @@ public static partial class JsonSerializer
         /// <exception cref="System.ArgumentNullException">
         /// <paramref name="utf8Json"/> is <see langword="null"/>.
         /// </exception>
-        public static IAsyncEnumerable<TValue> DeserializeAsyncEnumerable<[DynamicallyAccessedMembers(JsonHelpers.MembersAccessedOnRead)] TValue>(
+        public static IAsyncEnumerable<TValue?> DeserializeAsyncEnumerable<[DynamicallyAccessedMembers(JsonHelpers.MembersAccessedOnRead)] TValue>(
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs:201]: Good catch, thanks.

File: src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs (hunk 4 above)
Comment (layomia) [reply src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs:44]: Hm no, it isn't related to the new `DeserializeAsyncEnumerable` method. This diff is just refactoring the code to use a new `GetContentStream` method which contains logic that was previously duplicated, and would have been duplicated more in the new methods I'm adding. This particular method retains its call to the existing `DeserializeAsync`.

File: src/libraries/System.Net.Http.Json/tests/FunctionalTests/JsonContext/String.cs (hunk 5 above)
Comment (layomia) [reply src/libraries/System.Net.Http.Json/tests/FunctionalTests/JsonContext/String.cs:16]: Per offline chat, these are properties are defined in the same partial class.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/JsonContext/HighLowTemps.cs (hunk 7 above)
Comment (layomia) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/JsonContext/HighLowTemps.cs:58]: makes sense to update it. I've done so in this PR, but will update it in the generator in a follow up.

File: src/libraries/System.Net.Http.Json/tests/FunctionalTests/JsonContext/Person.cs (hunk 9 above)
Comment (layomia) [reply src/libraries/System.Net.Http.Json/tests/FunctionalTests/JsonContext/Person.cs:43]: It should never, unless we have a bug (e.g. a weird cross-pollination of options instances being used). The serializer should only pass the same context instance that the type info instance was a property on.

I'll think about this some more. We could be more defensive in the gen'd code and throw `InvalidOperationException` instead of leak `InvalidCastException`. 

To avoid any potential issues here entirely, we could make the method non-static and reference the context directly (instead of taking it as an argument), but I want to be sure of any perf implications first.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs (hunk 10 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs:92]: Just saw I only addressed this for the Serialize APIs. Will make follow up for Deserialize APIs.

**Comment by ericstj** [comment]
I think the backport is failing because the 2/3 PR hasn't been merged yet

**Review by tdykstra — COMMENTED** [review]
Everything in 2/3 and 3/3 LGTM, aside from one nit.

**Comment by danmoseley** [comment]
@tdykstra @carlossanlop I think it's great that docs folks are able to review the docs in the exact same PR that we have implementation and tests. That has to be so much more efficient long term.
